posix/devices/hostsock/hostsock.c

```diff
@@ -0,0 +1,957 @@
+        OE_RAISE_ERRNO(OE_EINVAL);
+    }
+
+done:
+
+    return ret;
+}
+
+static int _hostsock_setsockopt(
+    oe_fd_t* sock_,
+    int level,
+    int optname,
+    const void* optval,
+    oe_socklen_t optlen)
+{
+    int ret = -1;
+    sock_t* sock = _cast_sock(sock_);
+
+    oe_errno = 0;
+
+    if (!sock || !optval || !optlen)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (oe_posix_setsockopt_ocall(
+            &ret, sock->host_fd, level, optname, optval, optlen) != OE_OK)
+    {
+        OE_RAISE_ERRNO(OE_EINVAL);
+    }
+
+done:
+
+    return ret;
+}
+
+static int _hostsock_ioctl(oe_fd_t* sock_, unsigned long request, uint64_t arg)
+{
+    int ret = -1;
+    sock_t* sock = _cast_sock(sock_);
+
+    oe_errno = 0;
+
+    if (!sock)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (oe_posix_ioctl_ocall(&ret, sock->host_fd, request, arg) != OE_OK)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+done:
+
+    return ret;
+}
+
+static int _hostsock_getpeername(
+    oe_fd_t* sock_,
+    struct oe_sockaddr* addr,
+    oe_socklen_t* addrlen)
+{
+    int ret = -1;
+    sock_t* sock = _cast_sock(sock_);
+    oe_socklen_t addrlen_in = 0;
+
+    oe_errno = 0;
+
+    if (!sock)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (addrlen)
+        addrlen_in = *addrlen;
+
+    if (oe_posix_getpeername_ocall(
+            &ret,
+            sock->host_fd,
+            (struct oe_sockaddr*)addr,
+            addrlen_in,
+            addrlen) != OE_OK)
+    {
+        OE_RAISE_ERRNO(OE_EINVAL);
+    }
+
+done:
+
+    return ret;
+}
+
+static int _hostsock_getsockname(
+    oe_fd_t* sock_,
+    struct oe_sockaddr* addr,
+    oe_socklen_t* addrlen)
+{
+    int ret = -1;
+    sock_t* sock = _cast_sock(sock_);
+    oe_socklen_t addrlen_in = 0;
+
+    oe_errno = 0;
+
+    if (!sock)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (addrlen)
+        addrlen_in = *addrlen;
+
+    if (oe_posix_getsockname_ocall(
+            &ret,
+            sock->host_fd,
+            (struct oe_sockaddr*)addr,
+            addrlen_in,
+            addrlen) != OE_OK)
+    {
+        OE_RAISE_ERRNO(OE_EINVAL);
+    }
+
+done:
+
+    return ret;
+}
+
+static ssize_t _hostsock_read(oe_fd_t* sock_, void* buf, size_t count)
+{
+    return _hostsock_recv(sock_, buf, count, 0);
+}
+
+static ssize_t _hostsock_write(oe_fd_t* sock_, const void* buf, size_t count)
+{
+    return _hostsock_send(sock_, buf, count, 0);
+}
+
+static ssize_t _hostsock_readv(
+    oe_fd_t* desc,
+    const struct oe_iovec* iov,
+    int iovcnt)
+{
+    ssize_t ret = -1;
+    sock_t* sock = _cast_sock(desc);
+    void* buf = NULL;
+    size_t buf_size = 0;
+
+    if (!sock || (!iov && iovcnt) || iovcnt < 0 || iovcnt > OE_IOV_MAX)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* Flatten the IO vector into contiguous heap memory. */
+    if (oe_iov_pack(iov, iovcnt, &buf, &buf_size) != 0)
+        OE_RAISE_ERRNO(OE_ENOMEM);
+
+    /* Call the host. */
+    if (oe_posix_recvv_ocall(&ret, sock->host_fd, buf, iovcnt, buf_size) !=
```

Comment:
Need to check ret before the iov_sync

Reply:
Harmless but you suggestion improves efficiency so added.